### PR TITLE
Backfill

### PR DESCRIPTION
**Pull request by roidelapluie**
This pull request is a work in progress to add a command to turn graphite whisper database into TSDB blocks.

### CODE REVIEW

**Comment by roidelapluie** [comment]
Still need docs and has some performance issues, tackled at https://github.com/prometheus/prometheus/issues/8388 https://github.com/prometheus/prometheus/issues/8389

**Comment by roidelapluie** [comment]
Also still need to be able to use the mapper config file as the exporter does.

**Comment by roidelapluie** [comment]
I will rebase once #158 will be merged.

**Comment by roidelapluie** [comment]
@matthiasr This change is not compatible with AIX. What is your opinion here? Should be play with +build tags to maintain AIX compatibility or drop it?

**Comment by matthiasr** [comment]
Hmm, I would like to not drop anything we don't have to. Especially since
the whole point is legacy support … so if it's possible I would prefer to
keep AIX but if it's a lot of pain we can drop it.

**Comment by roidelapluie** [comment]
> Hmm, I would like to not drop anything we don't have to. Especially since the whole point is legacy support … so if it's possible I would prefer to keep AIX but if it's a lot of pain we can drop it.


TSDB is not AIX compliant. The reason is that AIX does not offer the ability to fsync directories, which can lead to data loss when TSDB renames/moves files. So, backfilling will not work in AIX.

An alternative is to use build flags so we do not offer that backfill feature on AIX, but the exporter still works as before.

**Comment by matthiasr** [comment]
That's what I thought! We don't need backfilling on AIX if Prometheus
doesn't run there, but it is good if the exporter itself still works.

**Comment by mdehollander** [comment]
Just a check if this PR is ready for use? Can I backfill whisher data to prometheus by:
- Creating TSDB blocks from the whisper directory with: `./getool create-blocks <whisper directory> [<output directory>]`
- Copy the data from the output directory to the prometheus data folder?

**Comment by matthiasr** [comment]
Oooh I completely missed that this is ready for review, I will take a look as soon as possible. @mdehollander please give this branch a spin and let us know how it goes for you!

**Comment by roidelapluie** [comment]
Yeah TBH the model of graphite makes that this is not efficient. I would need to probably use more lower-level to get better performances.

**Comment by matthiasr** [comment]
Slow is better than not at all! Let's get this in and optimize if necessary?

**Comment by matthiasr** [comment]
:yolo: I'll add a section to the README saying this is experimental. Let's see what comes out of that!

File: cmd/getool/main.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	app.HelpFlag.Short('h')
+
+	defaultDBPath := "data/"
+
+	importCmd := app.Command("create-blocks", "Import samples from OpenMetrics input and produce TSDB blocks. Please refer to the exporter docs for more details.")
+	// TODO(aSquare14): add flag to set default block duration
+	importFilePath := importCmd.Arg("whisper directory", "Directory of the whisper database.").Required().String()
+	importDBPath := importCmd.Arg("output directory", "Output directory for generated blocks.").Default(defaultDBPath).String()
+	importHumanReadable := importCmd.Flag("human-readable", "Print human readable values.").Short('r').Bool()
+	importBlockDuration := importCmd.Flag("block-duration", "TSDB block duration.").Default("2h").Duration()
```
Comment (matthiasr) [inline cmd/getool/main.go:37]: @roidelapluie I tried converting the test Whisper files into TSDB blocks; it took a long time and then ran out of file descriptors after 1023 blocks. It was much much quicker with 4096h blocks.

Given that the use case is importing long-term data, would there be a problem with defaulting to something much longer than 2h? I am writing the user documentation at the moment and "you must change this setting to get reasonable results" is not something I like writing.

**Pull request merged**